### ISSUE TITLE
Reproducible dataset order after shuffle from checkpoint

### ISSUE THREAD

**Issue by WolodjaZ** [tensorflow/tensorflow#58925]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Feature Request

### Source

source

### Tensorflow Version

tf 2.11

### Custom Code

No

### OS Platform and Distribution

macOS Monterey 12.6, Debian GNU/Linux bookworm/sid

### Python version

3.8.6

### CUDA/cuDNN version

11.4 (Debian)

### GPU model and memory

NVIDIA GeForce RTX 2080 Ti (Debian)

### Current Behaviour?

I would like to get reproducible results for datasets in which the data is shuffled. In PyTorch I could save rng and load rng to get the same shuffling order, but in Tensorflow I don't know how to do that. I would like to get results like this from a shuffled dataset. Additionally, I would also want to know how to do it on `Keras` datasets which are used in `model.fit()`

```shell
Iterate over dataset
  first batch of data [1,2,3,4,5]
save state of dataset

Iterate over dataset
  first batch of data [3,4,1,4,5]

load state of dataset
Iterate over dataset
  first batch of data [1,2,3,4,5]
```

### Standalone code to reproduce the issue

```python
# main imports
import random
import numpy as np
import tensorflow as tf

# set the number of threads for tensorflow explained later
tf.config.threading.set_intra_op_parallelism_threads(1)

def main():
    seed = 1
    random.seed(seed)
    tf.random.set_seed(seed)
    
    tf_dataset_train = (
        tf.data.Dataset.from_tensor_slices(tf.random.normal((4, 3)))
        .cache()
        .shuffle(4)
        .batch(2)
        .prefetch(tf.data.AUTOTUNE)
    )
    for i, batch in enumerate(tf_dataset_train.as_numpy_iterator()):
        if i == 0:
            print(f"First batch {batch}")
    
    # Save dataset shuffle state
    print("Next loading of the dataset")
    for i, batch in enumerate(tf_dataset_train.as_numpy_iterator()):
        if i == 0:
            print(f"First batch {batch}")
    
    # Load dataset shuffle state
    print("Get dataset order from first run")
    for i, batch in enumerate(tf_dataset_train.as_numpy_iterator()):
        if i == 0:
            print(f"First batch {batch}")


if __name__ == "__main__":
    main()
```

### Relevant log output

_No response_</details>

**Comment by SuryanarayanaY** [comment]
Hi @WolodjaZ ,

In Tensorflow we have Two Seeds one at Global level and another at Op level.When included both it shall create reproducible results.Please refer to the [source](https://www.tensorflow.org/api_docs/python/tf/random/set_seed) for examples on how it works. In your example you can set global seed using `tf.random.set_seed` and at Op level say for `shuffle()` you have `seed` argument where you can set some seed value here.

Iam attaching sample [gist](https://colab.research.google.com/gist/SuryanarayanaY/344582c1abd694c34e2b62d2c8c8906f/58925.ipynb) explaining how it works in Tensorflow.

Please refer and come back if more details needed.

Thank you!

**Comment by WolodjaZ** [comment]
Yes, thank you for your reply.

However, I had something else in mind. I would like to be able to take the
state from the Dataloader, save it and from the checkpoint resume with the
shuffle done in the same order. I was able to do this by passing a seed
that is changed every epoch for this init_seed + epoch function. However, I
don't think this is the best way to do it. Does the shuffle function use
this seed, or does it set a global seed every time? I want to do a
tutorial/blog on how to do repeatable results in several frameworks. The
code for this example is here
https://github.com/WolodjaZ/reproduc-ml-tutorial/blob/main/mnist_tensorflow.py,
if you have any advice I'd be happy to hear.

Sincerely,
Vladimir Zaigrajew

pon., 26 gru 2022 o 09:34 SuryanarayanaY ***@***.***>
napisał(a):

**Comment by WolodjaZ** [comment]
Hi SuryanarayanaY,

Your solution works well for accessing data from a specific epoch. By
giving the seed value to shuffle based on the epoch, I can easily access
the data from that epoch. However, I think that this solution may not be
perfect for handling data augmentation, where frameworks like PyTorch can
set all seeds for a given worker to handle this. In your framework, I am
not sure how to tackle this problem of reproducible data from datasets from
checkpoint.

Additionally, I have been trying to make reproducible training possible
from a checkpoint, both on CPU and GPU. Can you advise me on how to achieve
this? I have been using the code from this GitHub repository:
https://github.com/WolodjaZ/reproduc-ml-tutorial/blob/main/index.ipynb. I
want to be able to obtain the same loss and accuracy in the initial first
training loop and from the training loop starting from the second epoch
checkpoint. Specifically, after the first training loop, I want to be able
to continue training for a few more epochs and then come back to the second
epoch and continue training, with the aim of obtaining the same loss and
accuracy as in the initial training loop.  Thank you for your assistance.

Sincerely,
Vladimir Zaigrajew

śr., 4 sty 2023 o 12:48 SuryanarayanaY ***@***.***>
napisał(a):

**Comment by aaudiber** [comment]
The issue with the original repro is that it doesn't set the `reshuffle_each_iteration` argument to [shuffle](https://www.tensorflow.org/api_docs/python/tf/data/Dataset?version=nightly#shuffle). By default `shuffle` will reshuffle, so you get a different order each iteration.

You can [save dataset iterators into checkpoints](https://www.tensorflow.org/guide/checkpoint#create_the_checkpoint_objects), so that they can be restored to the exact state.

**Comment by WolodjaZ** [comment]
I apologise for the late reply.

Then I want to discuss a few things. First of all, I want to know more
about how checkpoints store an iterator where I will be comparing it with
my current approach (screenshot added in the appendix). So how the iterator
is stored, because if it is stored as an object (pickle for example) then
it is not optimal for storage space (mentioned: Note however that the
iterator checkpoints can be large). Compared to my solution, I only need to
store/know the seed value to recreate the batches. If the checkpoint only
stores seeds or something related optimised for disk space, then I would
try to implement it. Secondly, I understand how shuffle works and how the
reshuffle_each_iteration parameter affects, however, I want to shuffle my
train dataset every epoch and also want to know how shuffle shuffles this
dataset every time, so using this parameter doesn't help me (looking at my
example
https://github.com/WolodjaZ/reproduc-ml-tutorial/blob/main/mnist_tensorflow.py
).

wt., 17 sty 2023 o 23:23 Andrew Audibert ***@***.***>
napisał(a):

**Comment by SuryanarayanaY** [comment]
Hi @WolodjaZ ,

Apologies for the delay. Did you checked this [comment](https://github.com/tensorflow/tensorflow/issues/58925#issuecomment-1386163935) ? As mentioned the shuffle method has an argument `reshuffle_each_iteration` which is True by default. You need to set it to False explicitly that can probably fixes the issue.

Also I have gone through your repo and you have not set any thing to `reshuffle_each_iteration` in `shuffle()`.So by default its doing shuffling and hence changing the dataset order. Setting this argument to `False` might resolve your error.

Please check and confirm if still have problems?

Thank you!

**Comment by WolodjaZ** [comment]
@SuryanarayanaY, okay, I agree that it might work with `reshuffle_each_iteration`. However, what if I want to reshuffle after each iteration (epoch) and still be able to reproduce the iterator state (shuffle order) from the checkpoint? Is this scenario achievable?

@ujjwalnur, after quite some time, I stopped searching for a solution and settled with changing the seed every epoch and explicitly shuffling the dataset with it. Thanks to that approach, you can save the seed and reproduce the state of the iterator from the checkpoint. [The code I used](https://github.com/WolodjaZ/reproduc-ml-tutorial/blob/main/mnist_tensorflow.py)

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The shuffle function uses its own internal seed, which is independent of the global seed set by tf.random.set_seed. The combination of both the global and local seeds determines the final random number generation. The local seed changes with each epoch, but it still interacts with the global seed. [Ref](https://www.tensorflow.org/guide/checkpoint#create_the_checkpoint_objects)

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
